### PR TITLE
rustdoc: improve scroll locking in the rustdoc mobile sidebars

### PR DESCRIPTION
**Pull request by notriddle**
This PR prevents the main content area from scrolling while the mobile sidebar is open on documentation pages (porting the scroll locking behavior from the source sidebar to the regular sidebar), and also fixes some bad behavior where opening a "mobile" sidebar, and growing the viewport so that the "desktop" mode without scroll locking is activated, could potentially leave the page stuck.

This does not affect the behavior on larger screens. Only small ones, where the sidebar covers up the main content.

Split out from #98772

### CODE REVIEW

File: src/librustdoc/html/static/js/source-script.js
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/source-script.js:96]: Is it really useful to do that?

**Comment by GuillaumeGomez** [comment]
Since the code is almost to lock the scroll when the sidebar is open, we should maybe merge both codes. What do you think?

So for this, I'm a bit shared: it makes it somewhat coherent with the source sidebar but I don't think it's needed in this case. So better ask others what they think.

cc @Manishearth @jsha

**Comment by GuillaumeGomez** [comment]
So after reading [notriddle's comment](https://github.com/rust-lang/rust/pull/98772#issuecomment-1172633469), I think it's actually a good idea. Still double-checking there is no unforeseen UX issue but otherwise I agree with it.

File: src/librustdoc/html/static/js/source-script.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/source-script.js:96]: That's not my point. Is it useful to assume that users will resize their screen like this?

In any case, it can be handled in a simpler way: whether or not we're in "mobile width", apply the inline style when collapsing the sidebar. No need to add an event listener for "resize" with it.

**Comment by Manishearth** [comment]
I'm kinda +1 on this, but I defer to jsha

File: src/librustdoc/html/static/js/source-script.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/source-script.js:96]: I see. Then please add a GUI test reproducing this scenario.

File: src/test/rustdoc-gui/sidebar-mobile-scroll.goml
Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/sidebar-mobile-scroll.goml:26]: Indeed! Then let's just wait for @jsha's approval and let's go!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by jsha** [comment]
The scroll-lock trick uses a lot of JavaScript and seems complex. Also, I'm nervous about trying to avoid layout shifts by guessing the width of the scrollbar, which can vary based on OS.

I tried another approach, which rearranges which elements have a scroll bar. I think it works pretty well:

https://github.com/rust-lang/rust/compare/master...jsha:rust:scroll-fix?expand=1

> This fixes that by making the `<main>` element scrollable instead of `<body>`. Now the two scrollable elements (`<nav>` and `<main>`) are siblings rather than child and parent.

https://rustdoc.crud.net/jsha/scroll-fix/std/string/struct.String.html

This has the side effect that keyboard focus begins in the main content area rather than the sidebar, which seems more useful to me.

This also affects how the mobile topbar works, in a positive way. Instead of having it overlap the body and sidebar (in the z axis) with a corresponding offset in both, it is above the body and sidebar in the y axis. If we go this route we will probably need some slight tweaks in the docs.rs CSS to accomodate the topbar over there.

**Comment by jsha** [comment]
> Won't this cause https://github.com/rust-lang/docs.rs/issues/595 to come back?

Oh, interesting. I wasn't aware of that issue. Seems like yes, which would be a blocker.

**Comment by jsha** [comment]
@bors r+ rollup

Thanks for driving this forward!

**Pull request merged**
